### ISSUE TITLE
How to puclish from main thread

### ISSUE THREAD

**Issue by Fire7** [ClusterWS/ClusterWS#118]
### Submitting

- [ ] Bug
- [x] Question
- [ ] Suggestion
- [ ] Other

### Details 

Hello.
Is it possible to publish messages to channels from main thread (not workers)?
If so, how to do it?
I'm trying to find Broker or wss property in ClusterWS instance, but it is just empty object ({}).

**Comment by goriunov** [comment]
Hi @Fire7 it is not possible at right now, currently Publish is passed only to the workers and master is more like a load balancer. I don't see actual use case for that as all of the logic can be set in Worker processes and scaled automatically.

**Comment by carmas123** [comment]
Publish on main thread don't have sense...
However you can store a global client and use it to comunicate with other clients

**Comment by Fire7** [comment]
It make sense if I want to publish data from separated logic / thread / service. 
For example where are some stock-market and i want periodically to send data to clients subscribed to some channel, or some service triggered some event for several users and I want to notice them.
I think it is basic logic for microservice architecture.
Now it is only possible to create logic around each client action, but not use this library just as transport, what WebSockets are in essence.

**Comment by carmas123** [comment]
This is true but to do that you can do that with a "super client" connected to server after that the master server was started and use it to dispatch all messages to other "clients"...and I think that you problem is solved.

**Comment by Fire7** [comment]
Yes, but it is looks like crutch :)

**Comment by goriunov** [comment]
I think in this case it will be more useful to have a `in current worker publish function` (as event will be triggered periodically from the server ) and each worker will publish only to the users which are subscribed to the current worker process. I had some plans to add this.

@Fire7, @carmas123  What do you think about that ?

**Comment by carmas123** [comment]
Publish only on current worker is a bad idea for me...

**Comment by Fire7** [comment]
I'm not sure in internal realization of pub/sub system, but if I understand it right and it is same as in SocketCluster ( all workers connected to all brokers, and brokers divide channels among themselves by some rule(?)), good solution will be to publish from brokers itself (possibly it functionality is already implemented) and to set some logic in brokers (like Worker and like it is in SocketCluster). In this case all custom logic will be in broker (not main thread), but obviliously limitation here is only one broker for application.

Another good way I see is some API from main thread to manipulate brokers. It will be brilliant to get broker id by channel name from main thread and then call some method to publish, and it will solve all problems imho.

**Comment by Fire7** [comment]
I've just implement this solution in SocketCluster (with `sendToBroker` method), and it works well.
Unfortunately, where are no capability to get broker id by channel name out of the box. Single broker solve it.

**Comment by goriunov** [comment]
That is actually a good idea, alright i will consider to add this in version 4.0.0 (still need to get my head around it).

**Comment by goriunov** [comment]
@Fire7, @carmas123 I am working on implementing above logic.

@Fire7 some clarification about the `broker` `master` `worker` communication is that it works slightly different than in SocketCluster. Main difference is that brokers do not have devision between channels so you can publish to any broker and it will propagate to the right clients (we need multiple brokers to divide load between them with new v4 there is lots of improvement in communications).

So after checking all the things I think one of the ways which will not load master process and will simplify implementation is to run some function only on 0 Worker (each worker has it is own id from 0 -> ... number of workers) this will be a unique execution function which will have access to the publish method on that worker.

The main problem which i face is i don't have any idea yet how to call that function (param) which we should pass in to the ClusterWS.

What do you think about that ? Do you have any suggestions on the name ? Do you think is it better to run fn on master process ?
